### DISCUSSION TITLE
Use of enums vs prefix grouped constants

### DISCUSSION THREAD

**Discussion by scop** [home-assistant/architecture#548] [General]
I wonder if there's a reason we don't use enums more, but rather have lots and lots of separate (often str) constants just "grouped" by a prefix in their name? I suppose sometime it could have been that enums came in 3.4 only, and there were some problems with enum34, but is there anything these days why we should keep that general habit?

Enums constrain the set of values cleaner / "for real", document the possible values in function signatures, and play along with type hinting much better than separate constants.

Kind of a downside of enums would be that they can't be extended, but that I think is a solvable one, the fallback being just not using enums in such cases (but there can be better ways to deal with that and keep using enums still).

**Comment by thomasloven** [comment]
Do you have an example? Most string constants I've seen correspond to actual configuration keys for the yaml files.

**Reply by scop** [reply]
A lot homeassistant/const,py would serve as a fine example. Even if we'd exempt config key like constants from enumization (whicih I'm not sure why we'd do, they're a predefined set of str values anyway). Enumizing the config key constants wouldn't contribute much to making HA easier to code because they rarely occur in function signatures though. Anyway I think this is something we have going on all across the tree.

We're also failing to use defined constants here and there for a bunch of things and rather use magic ones for them. Use of enums would decrease the likelihood of this happening.

I'm a bit hesistant to post a specific example because it might not be the best one or might muddle the larger point. But meh, anyway the one that finally prompted me to start this discussion (have eyeballed this for a long time as something we should improve on) are the disabled_by values for config entries. There are constants for them in entity_registry.py that would work just fine as an enum, and having that it'd be immediately obvious what values to use for them in related functions. We're also failing to use those constants -- in this particular case core seems to be in a pretty good shape, but the test suite uses pretty much completely magic strings for them. I'm sure examples for which this happens in core too wouldn't be that hard to find.

**Reply by balloob** [reply]
Config entry state is a good example of what could have been an enum: 

https://github.com/home-assistant/core/blob/dev/homeassistant/config_entries.py#L67-L77

It's a fixed set and also documented as such:

https://developers.home-assistant.io/docs/config_entries_index#lifecycle

**Comment by balloob** [comment]
There is no reason.

**Comment by KapJI** [comment]
Data flow result type is another good candidate for enumization:
https://github.com/home-assistant/core/blob/dev/homeassistant/data_entry_flow.py#L16-L22

With current model adding a new value will require changing several places which check if value is from valid set:
https://github.com/home-assistant/core/blob/dev/homeassistant/data_entry_flow.py#L271-L279

With enums this can be easility generalized.

Enums should be better for typing and it's safer to check in the code if result has specific type from enum rather than comparing with some string.

Enums also force to use them explicitly. When string constants are expected, nothing stops from using plain string literals which may contain typos or just unsupported values.

Another disadvantage of string constants that they may be used incorrectly. Basically if 2 constants are equal they can be used interchangeably while underlying scopes can be very different which is not great for code readability.

The main disadvantage is that such refactorings will likely require changing a lot of code.

**Comment by frenck** [comment]
We have been moving to enums as much as possible already, IMHO this discussion can be closed now.

../Frenck
